plugins/security/nexus-ldap-plugin/src/main/java/org/sonatype/security/realms/ldap/internal/persist/DefaultLdapConfigurationManager.java

```diff
@@ -50,135 +50,120 @@
 {
   private final LdapConfigurationSource configurationSource;
 
-  private final LdapConfigurationValidator configurationValidator;
+  private final Validator validator;
 
   private final EventBus eventBus;
 
   private final SecuritySystem securitySystem;
 
   /**
-   * This will hold the current configuration in memory, to reload, will need to set this to null
+   * Configuration cache.
    */
-  private CLdapConfiguration ldapConfiguration = null;
+  private final LinkedHashMap<String, LdapConfiguration> cache;
+
+  private boolean cachePrimed;
 
   @Inject
   public DefaultLdapConfigurationManager(final LdapConfigurationSource configurationSource,
-                                         final LdapConfigurationValidator configurationValidator,
+                                         final Validator validator,
                                          final EventBus eventBus,
                                          final SecuritySystem securitySystem)
   {
     this.configurationSource = checkNotNull(configurationSource);
-    this.configurationValidator = checkNotNull(configurationValidator);
-    this.eventBus = eventBus;
+    this.validator = checkNotNull(validator);
+    this.eventBus = checkNotNull(eventBus);
     this.securitySystem = checkNotNull(securitySystem);
+    this.cache = Maps.newLinkedHashMap();
+    this.cachePrimed = false;
   }
 
   @Override
   public synchronized void clearCache() {
-    ldapConfiguration = null;
-    // fire event
+    cache.clear();
+    cachePrimed = false;
     eventBus.post(new LdapClearCacheEvent(this));
   }
 
   @Override
   public synchronized void setServerOrder(final List<String> orderdServerIds)
-      throws InvalidConfigurationException
+      throws IllegalArgumentException
   {
-    final List<CLdapServerConfiguration> ldapServers = getConfiguration().getServers();
-    final ValidationResponse vr = configurationValidator.validateLdapServerOrder(ldapServers, orderdServerIds);
-    if (vr.getValidationErrors().size() > 0) {
-      throw new InvalidConfigurationException(vr);
-    }
-
-    // build a map so its easier
-    final Map<String, CLdapServerConfiguration> idToServerMap = Maps.newHashMap();
-    for (CLdapServerConfiguration ldapServer : ldapServers) {
-      idToServerMap.put(ldapServer.getId(), ldapServer);
-    }
-    // now reorder them
-    final List<CLdapServerConfiguration> newOrderedldapServers = Lists.newArrayList();
-    for (String serverId : orderdServerIds) {
-      newOrderedldapServers.add(idToServerMap.get(serverId));
+    checkNotNull(orderdServerIds);
+    final LinkedHashMap<String, LdapConfiguration> configuration = getConfiguration();
+    final Set<String> newOrder = Sets.newHashSet(orderdServerIds);
```

Comment:
I think this is just checking that the ordered list contain the same ids (ie. no more, no less) that are in the configuration. If so then order here doesn't matter - but then the check and the error message could be improved to make this clear.